Sources/PackageModel/Target.swift

```diff
@@ -356,6 +356,9 @@ public class ClangTarget: Target {
 
     /// The path to include directory.
     public let includeDir: AbsolutePath
+    
+    /// The target's module map type; determined by the file system layout, and controls whether the build system will use a custom module map, generate a module map, etc.
+    public let moduleMapType: ModuleMapType
```

Comment:
I am not really sure that it makes sense to encode the module map type in the target. Clients which need to find out the module map type can directly call ModuleMapGenerator API, right?

Reply:
I think it makes sense to separate the decision from the generator, so clients could provide their own generator if desired, but could still use the type as determined by libSwiftPM.

Reply:
Right, the longer-term goal here is to have the information vended by libSwiftPM include all the semantic information that's relevant about the target.  One of those is the disposition of its headers.  It still makes sense to have a module map generation facility in SwiftPM (at least while it has its own custom build system), but I expect that many clients that can already generate module maps would want to use their own facility and just needs to know the decisions that SwiftPM made.  This can also include things like showing this information to users in some form, e.g. when inspecting a package-defined target to find out how SwiftPM interpreted it.  The plan is also to add this to the `describe` command.

As I mentioned in the previous response, this PR has the minimal support that could be pulled back to 5.3, but as it is probably too late for that kind of change in 5.3, I've converted this to a draft PR will I fill in the rest of the changes to complete the picture here.  I'll probably still keep them as separate independently testable commits, for clarity.

Reply:
This looks great, this is pretty much exactly what I was envisioning w.r.t 50872333 👍